modules/distributor/queue/queue.go

```diff
@@ -0,0 +1,224 @@
+package queue
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/weaveworks/common/user"
+	"go.uber.org/atomic"
+)
+
+type request[T any] struct {
+	data T
+}
+
+type ProcessFunc[T any] func(ctx context.Context, data T) error
+
+// Queue represents a single tenant's queue.
+type Queue[T any] struct {
+	// wg is used to wait for the workers to drain the queue while stopping
+	wg sync.WaitGroup
+
+	logger log.Logger
+
+	name           string
+	tenantID       string
+	workerCount    int
+	size           int
+	reqChan        chan *request[T]
+	fn             ProcessFunc[T]
+	workersCloseCh chan struct{}
+
+	pushesTotalMetrics         *prometheus.CounterVec
+	pushesFailuresTotalMetrics *prometheus.CounterVec
+	lengthMetric               *prometheus.GaugeVec
+
+	readOnly *atomic.Bool
+}
+
+func New[T any](cfg Config, logger log.Logger, reg prometheus.Registerer, fn ProcessFunc[T]) *Queue[T] {
+	pushesTotalMetrics := prometheus.NewCounterVec(prometheus.CounterOpts{
+		Namespace: "tempo",
+		Subsystem: "distributor",
+		Name:      "queue_pushes_total",
+		Help:      "Total number of successful requests queued up for a tenant to the generatorForwarder",
+	}, []string{"name", "tenant"})
+	pushesFailuresTotalMetric := prometheus.NewCounterVec(prometheus.CounterOpts{
+		Namespace: "tempo",
+		Subsystem: "distributor",
+		Name:      "queue_pushes_failures_total",
+		Help:      "Total number of failed pushes to the queue for a tenant to the generatorForwarder",
+	}, []string{"name", "tenant"})
+	lengthMetric := prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Namespace: "tempo",
+		Subsystem: "distributor",
+		Name:      "queue_length",
+		Help:      "Number of queued requests for a tenant",
+	}, []string{"name", "tenant"})
+
+	var alreadyRegisteredErr prometheus.AlreadyRegisteredError
+	err := reg.Register(pushesTotalMetrics)
+	if err != nil && !errors.As(err, &alreadyRegisteredErr) {
+		_ = level.Warn(logger).Log("msg", "failed to register queue_pushes_total metric", "err", err)
+	}
+	err = reg.Register(pushesFailuresTotalMetric)
+	if err != nil && !errors.As(err, &alreadyRegisteredErr) {
+		_ = level.Warn(logger).Log("msg", "failed to register queue_pushes_failures_total metric", "err", err)
+	}
+
+	err = reg.Register(lengthMetric)
+	if err != nil && !errors.As(err, &alreadyRegisteredErr) {
+		_ = level.Warn(logger).Log("msg", "failed to register queue_length metric", "err", err)
+	}
+
+	return &Queue[T]{
```

Comment:
we lost `workerAliveCount` from the previous implementation. it's not clear to me what the purpose of that var was. @kvrhdn do you see any issues with removing it?

Reply:
Workers can panic and that wouldn't get propagated since they're running in goroutines, so we could end up in a situation where workers are exhausted but they're not marked as such. The idea was to detect this and spawn new workers or reset the queue.

Reply:
Hmm, interesting. Interesting solution, however, I think that there were a couple of concurrency issues with it (hence the TODO there). Maybe we could create a separate issue for this one and leave it as is for now? I am also fine with bringing back the `workerAliveCount`, so if you prefer that, let me know.

Reply:
I'm fine with removing it too 👍

Reply:
We used `workerAliveCount` to detect when a worker go routine panics and crashes. `shouldUpdate` will check `workerAliveCount` with the desired count and recreate the workers if one or more is missing.
We had to keep a separate counter because `WaitGroup` doesn't say how many workers are active.

So the risk here is that multiple workers can crash, we won't detect this and the queue will struggle to keep up.

That said, I'm not sure if we ever saw this happen... I'm fine with dropping the code, if we end up having performance issues we should see the panics in the logs and can add extra safeguards.